### PR TITLE
Direct namespace match (AO-17676)

### PR DESCRIPTION
**Pull request by adamiklukasz**

### CODE REVIEW

File: v2/internal/util/metrictree/directmatch_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		shouldMatch: true,
+	},
+	{
+		ns:          "|log-files|[file=/tmp/sample1.log]|string_line",
+		filter:      "|log-files|[file=/tmp/sample.log]|string_line",
+		shouldMatch: false,
+	},
+	{
+		ns:          "/syslog/[ip_address=127.0.0.1]/[hostname=localhost]/string_line",
+		filter:      "/syslog/*/[hostname=localhost]/string_line",
```
Comment (lootek) [inline v2/internal/util/metrictree/directmatch_test.go:88]: please add a test case (with the same `ns`):

```suggestion
		filter:      "/syslog/**/string_line",
```

File: v2/internal/util/metrictree/directmatch_test.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		shouldMatch: true,
+	},
+	{
+		ns:          "/kubernetes/[pod=io1453]/cpu",
+		filter:      "/kubernetes/[pod=io1453]",
+		shouldMatch: true,
+	},
+	{
+		ns:          "/kubernetes/[pod=io1453]/cpu",
+		filter:      "/kubernetes/[pod]",
```
Comment (lootek) [inline v2/internal/util/metrictree/directmatch_test.go:48]: please add a test case (with the same ns):

```suggestion
		filter:      "/kubernetes/[node]",
```

File: v2/internal/util/metrictree/directmatch_test.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		shouldMatch: false,
+	},
+	{
+		ns:          "/mock/metric/c",
+		filter:      "/another-plugin",
+		shouldMatch: false,
+	},
+	{
+		ns:          "/mock/metric/c",
+		filter:      "/",
```
Comment (lootek) [inline v2/internal/util/metrictree/directmatch_test.go:114]: please add test cases (with the same ns):

```suggestion
		filter:      "/*",
```

```suggestion
		filter:      "/**",
```

File: v2/internal/util/metrictree/directmatch_test.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		fmt.Printf("i=%#v\n", i)
+		match, err := MatchNsToFilter(te.ns, te.filter)
+
+		if te.shouldMatch {
+			require.True(t, match, "Namespace should match to a given filter")
+		} else {
+			require.False(t, match, "Namespace shouldn't match to a given filter")
+		}
+
+		if !te.expectError {
```
Comment (lootek) [inline v2/internal/util/metrictree/directmatch_test.go:130]: please invert condition for consistency with the above

File: v2/internal/util/metrictree/namespace.go [hunk 5]
```diff
@@ -194,7 +194,7 @@ func newStaticAnyElement() *staticAnyElement {
 }
 
 func (*staticAnyElement) Match(s string) bool {
-	return isValidIdentifier(s)
+	return s != "*" && s != "**" && isValidGroupIdentifier(s)
```
Comment (lootek) [inline v2/internal/util/metrictree/namespace.go:197]: I'm ok with this change but let's wait for @adamiklukasz to confirm if it won't have some side effects in other places

File: v2/internal/util/metrictree/directmatch_test.go (hunk 1 above)
Comment (tomekkolo) [reply v2/internal/util/metrictree/directmatch_test.go:88]: done

Comment (tomekkolo) [reply v2/internal/util/metrictree/directmatch_test.go:88]: ** cannot be in the middle, it is for last element

File: v2/internal/util/metrictree/directmatch_test.go (hunk 2 above)
Comment (tomekkolo) [reply v2/internal/util/metrictree/directmatch_test.go:48]: done

File: v2/internal/util/metrictree/directmatch_test.go (hunk 3 above)
Comment (tomekkolo) [reply v2/internal/util/metrictree/directmatch_test.go:114]: done

File: v2/internal/util/metrictree/directmatch_test.go (hunk 1 above)
Comment (lootek) [reply v2/internal/util/metrictree/directmatch_test.go:93]: ah, ok

File: v2/internal/util/metrictree/directmatch_test.go [hunk 6]
```diff
@@ -88,6 +93,16 @@ var testCases = []testCase{
 		filter:      "/syslog/*/[hostname=localhost]/string_line",
 		shouldMatch: true,
 	},
+	{
+		ns:          "/syslog/[ip_address=127.0.0.1]/[hostname=localhost]/string_line",
+		filter:      "/syslog/[ip_address=127.0.0.1]/[hostname=localhost]/string_line",
```
Comment (lootek) [inline v2/internal/util/metrictree/directmatch_test.go:98]: ?

Comment (tomekkolo) [reply v2/internal/util/metrictree/directmatch_test.go:98]: check if same are match :)

**Review by lootek — APPROVED** [review]

**Review by tomekkolo — APPROVED** [review]

File: v2/internal/util/metrictree/directmatch.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+	// handle special case when filter have 1 element
+	el, _, err := SplitNamespace(filter)
+	if err != nil {
+		return false, fmt.Errorf("invalid filter: %w", err)
+	}
+
+	var parsedFilter *Namespace
+	if len(el) == 2 { // el[0] is empty
+		if el[1] == "" || el[1] == "*" || el[1] == "**" {
```
Comment (adamiklukasz) [inline v2/internal/util/metrictree/directmatch.go:41]: there are consts in 	parser.go in the same module

Comment (tomekkolo) [reply v2/internal/util/metrictree/directmatch.go]: done

**Review by tomekkolo — APPROVED** [review]

**Review by lootek — APPROVED** [review]

**Review by lootek — APPROVED** [review]

**Pull request merged**